Working splash screen in Linux GameLauncher

Fireronin opened the pull request:
## What does this PR do?

This PR allows for setting of splash screen that is rendered using xcb, therefore it removes black screen while renderer is working.

Splash screen image is passed via xxd generated .h file during compilation, from png file named logo.png

The use of generic malloc for data vector and not AZ vectors is due to the fact that functions like xcb_destroy_image, expect ability to deallocate vector with data.

It adds dependency on libxcb-image0-dev


## How was this PR tested?

It was tested in debug builds, of project.

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +243,127 @@ namespace AzFramework
+        AZ::SettingsRegistryInterface::FixedValueString splashScreenImagePath;
+        AZ::SettingsRegistryInterface::FixedValueString assetCachePath;
+        static const AZStd::string SplashLogoSetregPath = "/O3DE/xcb/SplashScreenImagePath";
+        static const AZStd::string AssetsSetregPath = "/O3DE/Runtime/FilePaths/CacheProjectRootFolder";
```
You cannot have a cache folder predefined in the registry, since it is different with GameLauncher run in install target and YourProject.GameLauncher. Also it will be different for different platforms. I can see you need to get a global path for XCB, but please use a different approach.

Fireronin replied:
I currently don't have idea how to do this in clean way so I added this logo binary data using .h header, this still looks ok and header file is only 1.3MB, tho size of the binary data in exe should be much smaller around 300-400KB

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +243,127 @@ namespace AzFramework
+
+        if (!(settingsRegistry && settingsRegistry->Get(splashScreenImagePath, SplashLogoSetregPath.c_str())))
+        {
+            printf("SplashScreenImagePath not found\n");
```
Why not AZ_Printf?

michalpelka replied:
Or AZ_Warning, or AZ_Assert?

Fireronin replied:
Adressed

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +243,127 @@ namespace AzFramework
+            return;
+        }
+        const AZStd::string splashScreenImagePath2 =
+            AZStd::string::format("%s/linux/%s", assetCachePath.c_str(), splashScreenImagePath.c_str());
```
Use AZ::IO::Path here. And do not hardcode the platform name.

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +243,127 @@ namespace AzFramework
+        png_read_image(png_ptr, row_pointers);
+        fclose(fp);
+        unsigned int stride = width * 4;
+        uint8_t* data = (uint8_t*)malloc(height * stride);
```
Use smth from C++ not C.

michalpelka replied:
Whole PNG code should be contained in a separated method/function.

Fireronin replied:
As explained this malloc is needed as xcb expects ability to free data, variable use of vector results in heap corruption.

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -439,4 +590,4 @@ namespace AzFramework
         }
     }
-} // namespace AzFramework
+} // namespace AzFramework
```
```suggestion
} // namespace AzFramework

```

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -68,15 +79,26 @@ namespace AzFramework
-            m_xcbConnection, XCB_COPY_FROM_PARENT, m_xcbWindow, xcbParentWindow, aznumeric_cast<int16_t>(geometry.m_posX),
-            aznumeric_cast<int16_t>(geometry.m_posY), aznumeric_cast<int16_t>(geometry.m_width), aznumeric_cast<int16_t>(geometry.m_height),
-            borderWidth, XCB_WINDOW_CLASS_INPUT_OUTPUT, m_xcbRootScreen->root_visual, eventMask, valueList);
+            m_xcbConnection,
```
Please revert formatting only-changes. It makes review hard.

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -6,13 +6,18 @@
+#include <AzCore/Settings/SettingsRegistry.h>
+#include <AzCore/std/containers/vector.h>
 #include <AzFramework/Application/Application.h>
+#include <AzFramework/Asset/AssetSystemBus.h>
```
Unused include. I think it is not needed?

Fireronin replied:
Good catch

michalpelka reviewed (changes requested):
I cannot accept a change, since it needs a hardcoded location of the cache folder concerning executable. It, most probably, works only with developers' setup when the GameLauncher target is built. I would suggest testing the monolithic build and trying to run the install target.
Refer to https://development--o3deorg.netlify.app/docs/user-guide/interactivity/robotics/deploying-simulation/  https://development--o3deorg.netlify.app/docs/user-guide/packaging/

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +243,127 @@ namespace AzFramework
+                        x->width,
+                        x->height);
+                    xcb_flush(m_xcbConnection);
+                    goto end;
```
https://stackoverflow.com/questions/3517726/what-is-wrong-with-using-goto

spham-amzn reviewed (changes requested):
We cannot use the approach of checking in the splash screen as binary data for the splash screen since the splash screen will be specific to the launcher. It needs to be an image that is specific to the project itself, so there probably needs a mechanism to do this that is outside of the asset processor itself during the build to generate the splash screen image as part of the build process.

Fireronin commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+        }
+
+        xcb_free_pixmap(m_xcbConnection, pixmap);
+    }
```
Hi, no it's not missing, the xcb_image_destroy is responsible for freeing the data vector, also this library is not aware of AZStd, so I wanted to use standard malloc to not have memory leak, above I'm using free, because the "ownership" was not yet transferred to xcb

PawelLiberadzki commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+
+        auto get_image_data_static = [](int &width, int &height) -> uint8_t*
+        {
+            width = O3DE_LOGO::width;
+            height = O3DE_LOGO::height;
```
Both are narrowing implicit casts that raise warnings.
```suggestion
            width = static_cast<int>(O3DE_LOGO::width);
            height = static_cast<int>(O3DE_LOGO::height);
```
Or you can change width and height (and m_width and m_height) to type best matching what is used in xcb calls.

PawelLiberadzki commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+            
+            for (int i = 0; i < width * height; i++)
+            {
+                logo_binary_data[i * 4 + 0] = O3DE_LOGO::header_data_cmap[(unsigned char)O3DE_LOGO::header_data[i]][0];
+                logo_binary_data[i * 4 + 1] = O3DE_LOGO::header_data_cmap[(unsigned char)O3DE_LOGO::header_data[i]][1];
+                logo_binary_data[i * 4 + 2] = O3DE_LOGO::header_data_cmap[(unsigned char)O3DE_LOGO::header_data[i]][2];
```
No need to cast to unsigned char in these lines?

PawelLiberadzki commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+            free(data);
+            return;
+        }
+        // This uses hardcoded 8bit depth , TODO check on HDR monitor
```
Did this TODO maybe take place?

michalpelka replied:
Open issue when PR will be merged. Delete Todo from code @Fireronin .

PawelLiberadzki commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+                drawn = true;
+                break;
+            }
+            break;
```
This break seems redundant to the one above.

PawelLiberadzki commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+        int width = 0;
+        int height = 0;
+
+        auto get_image_data_static = [](int &width, int &height) -> uint8_t*
```
I would stick to one naming convention, which seems to be camelCase in most cases in this code base.

PawelLiberadzki replied:
This is also a lambda used only once, below. Maybe its OK to used this code directly?

Fireronin replied:
This lambda pattern might be useful if someone will want to create some other strategy of loading image, and I don't think it creates a sensible division of code, for part that loads the image and part that displays it

spham-amzn replied:
I don't think thats a practical reason to use a lambda for this purpose. If another strategy is applied, is it expected to create another lambda here?

michalpelka replied:
Please keep it simple

Fireronin replied:
Lambda removed

PawelLiberadzki commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+        }
+        // This uses hardcoded 8bit depth , TODO check on HDR monitor
+        xcb_image_t* image =
+            xcb_image_create_native(m_xcbConnection, width, height, XCB_IMAGE_FORMAT_Z_PIXMAP, 24, data, width * height * 4, data);
```
As it was mentioned above:
```suggestion
            xcb_image_create_native(m_xcbConnection, width, height, XCB_IMAGE_FORMAT_Z_PIXMAP, 24, data, width * height * depth, data);
```
Please consider similar approach with 24 here.

PawelLiberadzki commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+        xcb_void_cookie_t cookie;
+        xcb_pixmap_t pixmap = xcb_generate_id(m_xcbConnection);
+        cookie = xcb_create_pixmap_checked(m_xcbConnection, m_xcbRootScreen->root_depth, pixmap, m_xcbWindow, width, height);
+        if (xcb_generic_error_t* error = xcb_request_check(m_xcbConnection, cookie))
```
It looks like you are able to do this checkout without allocating data (line 255 above). It could save some resources (and in some cases) if it would be possible to move malloc after this if-statement and read only the necessary width and height.

Fireronin replied:
Yaa, but width and height are computed during loading of the image, I like that all image data is computed at the same time

Fireronin commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,93 @@ namespace AzFramework
+        {
+            AZ_Printf("XCB window init", "Error in xcb_create_pixmap: %d\n", error->error_code);
+            free(error);
+            free(data);
```
Used for parity with xcb normal malloc

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/azframework_xcb_files.cmake:
```diff
@@ -6,6 +6,10 @@
 #
 
+# generate logo.h, based on logo.png
+execute_process(COMMAND xxd -i logo.png logo.h
+                WORKING_DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/AzFramework)
```
This still does not alleviate the concern of having a project-specific splash screen. The logo.png is still in the engine source. This approach also will write the generated `logo.h` file back to `AzFramework` path, and that by itself raises a some issues:
* The folder is under git control, so the file will be marked as a new file
* For installer builds, it generates a file into the installation files, which generally is not preferred
* For Linux SNAP packages, this will fail because the engine folder is in an immutable folder structure, and therefore cannot be written anyways.

michalpelka replied:
Also `xxd` needs to be in prerequisites for linux build.

Fireronin replied:
Thank you for pointing those issues, I think now I resolved them, I'm using cmake to generare this splash file, from Splash.png located in Project resources folder, to build directory, so no git controlled folder is touched, also it should solve with SNAP. 
This approach also allows for per project logo specification as it takes data from project files

spham-amzn commented on cmake/Platform/Linux/Packaging_linux.cmake:
```diff
@@ -48,6 +48,7 @@ elseif("$ENV{O3DE_PACKAGE_TYPE}" STREQUAL "DEB")
         libxkbcommon-dev                        # For xcb keyboard input
         libxcb-xfixes0-dev                      # For mouse input
         libxcb-xinput-dev                       # For mouse input
+        libxcb-image0-dev                       # For xcb image support
```
If we are adding another xcb package dependency, then it also needs to be added to the build node initialization list, i.e. https://github.com/o3de/o3de/blob/development/scripts/build/build_node/Platform/Linux/package-list.ubuntu-jammy.txt . 
And also in order for AR to run successfully, the linux nodes will need to be rebuilt first with the new nodes, which means that change may need a separate PR to go in first, and then coordination with sig-build is needed to propogate the changes.

michalpelka replied:
@Fireronin could you prepare a separate PR as @spham-amzn proposed and link here.

Fireronin replied:
Here is pr you asked for

https://github.com/o3de/o3de/pull/17925

spham-amzn replied:
This will also require an update the documentation [here](https://github.com/o3de/o3de.org/blob/5291e85d708ff2955c43cddd52dc918b6c27d1da/content/docs/welcome-guide/requirements.md?plain=1#L160-L182):

spham-amzn replied:
Also note that you will need to work with sig-build to get the changes to the AR jenkins nodes from https://github.com/o3de/o3de/pull/17925 before this change, otherwise it will not pass AR, unless you make the splash screen optional (See the additional comment in this review)

spham-amzn commented:
I think the logic to process the splash screen for the launcher belongs in the [LauncherUnified](https://github.com/o3de/o3de/tree/development/Code/LauncherUnified) project and not in AzFramework.  The generation of the logo.h should be based on a png file that exists in the game project itself, and that file should go into the intermediate build folder with the include path appended instead of placing it directly in AzFramework.

Fireronin commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,142 @@ namespace AzFramework
+        int height = 0;
+        const unsigned int bytesPerPixel = 4;
+
+        auto get_image_static_png = [](int& width, int& height) -> uint8_t* {
```
Sure I can remove it

lemonade-dm commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,135 @@ namespace AzFramework
+            png_size_t read_length)
+        {
+            ReadDataHandle* handle = (ReadDataHandle*)png_get_io_ptr(png_ptr);
+            const png_byte* png_src = handle->data.data + handle->offset;
+
+            memcpy(raw_data, png_src, read_length);
+            handle->offset += read_length;
+        };
+
+        png_structp png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+        png_infop info_ptr = png_create_info_struct(png_ptr);
+
+        ReadDataHandle handle = {{ logo_png, logo_png_len},0};
+
+        png_set_read_fn(png_ptr, &handle, ReadDataFromInputStream);
+        png_read_info(png_ptr, info_ptr);
+
+        int width = png_get_image_width(png_ptr, info_ptr);
+        int height = png_get_image_height(png_ptr, info_ptr);
+
+        png_bytep row_pointers[height];
```
Per the coding standard variable names are required to be [lowerCamelCase](https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md#variable-names).

Fireronin replied:
good to know :smile:

lemonade-dm commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +229,135 @@ namespace AzFramework
+            png_size_t read_length)
+        {
+            ReadDataHandle* handle = (ReadDataHandle*)png_get_io_ptr(png_ptr);
+            const png_byte* png_src = handle->data.data + handle->offset;
+
+            memcpy(raw_data, png_src, read_length);
+            handle->offset += read_length;
+        };
+
+        png_structp png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+        png_infop info_ptr = png_create_info_struct(png_ptr);
+
+        ReadDataHandle handle = {{ logo_png, logo_png_len},0};
+
+        png_set_read_fn(png_ptr, &handle, ReadDataFromInputStream);
+        png_read_info(png_ptr, info_ptr);
+
+        int width = png_get_image_width(png_ptr, info_ptr);
+        int height = png_get_image_height(png_ptr, info_ptr);
+
+        png_bytep row_pointers[height];
```
The height variable is not a constant and therefore can't be used as the size field of an array as that is a compile time calculation.
https://godbolt.org/z/bG7rrb1bf

However the rowPointers variable can use the array version of  `unique_ptr` per the C++ [Best Practices Guide](https://github.com/o3de/sig-core/blob/main/governance/C%2B%2B-Best-Practices-Guide.md#smart-pointers).
```suggestion
            const png_byte* png_src = handle->data.data + handle->offset;

            memcpy(raw_data, png_src, read_length);
            handle->offset += read_length;
        };

        png_structp png_ptr = png_create_read_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
        png_infop info_ptr = png_create_info_struct(png_ptr);

        ReadDataHandle handle = {{ logo_png, logo_png_len},0};

        png_set_read_fn(png_ptr, &handle, ReadDataFromInputStream);
        png_read_info(png_ptr, info_ptr);

        int width = png_get_image_width(png_ptr, info_ptr);
        int height = png_get_image_height(png_ptr, info_ptr);

        AZStd::unique_ptr<png_bytep[]> rowPointers;
```

If the allocated memory needs to be handed over to the `libpng` that can be done via using the [unique_ptr::release](https://en.cppreference.com/w/cpp/memory/unique_ptr/release) function, but looking at the code I don't see any place where a `libpng` deallocates the memory.

Fireronin replied:
Thank you for the help, changes applied

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.h:
```diff
@@ -55,9 +55,14 @@ namespace AzFramework
         void InitializeAtoms();
         void GetWMStates();
 
+#ifdef SPLASH_IMAGE_EXISTS
```
I would left draw splash function. I would #ifdef SPLASH_IMAGE_EXISTS only used inside DrawSplash implementation. We should use as little macros as possible.

Fireronin replied:
But on the other hand this leaves a function that is unimplemented, which isn't much better

michalpelka commented on Code/Framework/AzFramework/Platform/Common/Xcb/azframework_xcb_files.cmake:
```diff
@@ -6,6 +6,17 @@
+    execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_BINARY_DIR}/AzFramework)
+    file(REMOVE ${CMAKE_BINARY_DIR}/AzFramework/Splash.h)
+    add_definitions(-DSPLASH_IMAGE_EXISTS)
+    execute_process(COMMAND xxd -i Splash.png ${CMAKE_BINARY_DIR}/AzFramework/Splash.h
```
Splash.png is hardcoded, there is no way for the user to change it after compiling the engine.

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+    // otherwise it will return without doing anything
+    void XcbNativeWindow::DrawSplash()
+    {
+#ifndef SPLASH_IMAGE_EXISTS
+    return;  // failed to find splash image no need to draw
+#else
```
* You can simply use `#if` instead of `ifdef/ifndef`. This works for situations where `SPLASH_IMAGE_EXISTS` is defined or not, and if its defined, it will evaluate the value (true/false or 1/0)
* Seems unnecessary for both the `return` statement and the inverted #ifndef block here. The implementation of this function could simply be wrapped around a `#if SPLASH_IMAGE_EXISTS`/`#endif` block

Fireronin replied:
Oh you are right this makes a lot of sense

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+#ifndef SPLASH_IMAGE_EXISTS
+    return;  // failed to find splash image no need to draw
+#else
+        const unsigned int bytesPerPixel = 4;
```
```suggestion
        constexpr unsigned int bytesPerPixel = 4;
```

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+#else
+        const unsigned int bytesPerPixel = 4;
+        
+        typedef struct
+        {
+            const png_byte* data;
+            const png_size_t size;
+        } DataHandle;
```
No need for `typedef struct ..`. This is a c++ source file, you can declare this using c++ syntax.

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+            const png_size_t size;
+        } DataHandle;
+
+        typedef struct
+        {
+            const DataHandle data;
+            png_size_t offset;
+        } ReadDataHandle;
```
Same comment here

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+            png_byte* rawData,
+            png_size_t readLength)
+        {
+            ReadDataHandle* handle = (ReadDataHandle*)png_get_io_ptr(pngPtr);
```
use C++ style casting
```suggestion
            ReadDataHandle* handle = static_cast<ReadDataHandle*>(png_get_io_ptr(pngPtr));
```

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+        int height = png_get_image_height(pngPointer, pngInfoPointer);
+
+        AZStd::unique_ptr<png_bytep[]> rowPointers = AZStd::make_unique<png_bytep[]>(height);
+        //png_bytep rowPointers[height];
```
Remove commented out code

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/azframework_xcb_files.cmake:
```diff
@@ -6,6 +6,43 @@
 #
 #
 
+# If not set explicitly, look for a splash image in the project directory
+if(NOT DEFINED SPLASH_FILE)
+    # start by looking for file in the project directory
+    if(EXISTS ${PROJECT_SOURCE_DIR}/Resources/Splash.png)
+        set(SPLASH_FILE ${PROJECT_SOURCE_DIR}/Resources/Splash.png)
+    endif()
+    # if not found, assume it's engine centric build
+    if(DEFINED LY_PROJECTS)
+        if(EXISTS ${LY_PROJECTS}/Resources/Splash.png)
+            set(SPLASH_FILE ${LY_PROJECTS}/Resources/Splash.png)
+        endif()
+    endif ()
+endif()
+# If value is set, check if it exists and generate the header file
+if(DEFINED SPLASH_FILE)
+    file(TO_CMAKE_PATH ${SPLASH_FILE} SPLASH_FILE_RESOLVED_PATH)
+    # check if it's a png file
+    if(NOT ${SPLASH_FILE_RESOLVED_PATH} MATCHES ".*\\.png")
+        message(WARNING "Splash file must be a PNG image, omitting splash screen generation.")
+    else()
+        if(EXISTS ${SPLASH_FILE_RESOLVED_PATH})
+            execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_BINARY_DIR}/AzFramework)
+            execute_process(COMMAND ${CMAKE_COMMAND} -E copy ${SPLASH_FILE_RESOLVED_PATH} ${CMAKE_BINARY_DIR}/AzFramework/Splash.png)
+            file(REMOVE ${CMAKE_BINARY_DIR}/AzFramework/Splash.h)
+            add_definitions(-DSPLASH_IMAGE_EXISTS)
+            execute_process(COMMAND xxd -i Splash.png Splash.h
+                    WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/AzFramework
+            )
+            include_directories(${CMAKE_BINARY_DIR}/AzFramework)
+        else ()
+            message(WARNING "${SPLASH_FILE} not found, no splash screen will be displayed"
+                    " (set SPLASH_FILE to a valid path to a splash image to display it)")
+        endif()
+    endif()
+endif()
+
+
```
This is not the right place to put this logic. This needs to be in a separate cmake file. Then the separate cmake file can be included in the "xcb" if block in `platform_nativeui_linux.cmake`

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+            switch (event->response_type & ~0x80)
+            {
+            case XCB_EXPOSE: {
+                xcb_expose_event_t* x = (xcb_expose_event_t*)event;
```
nit: Can we use something more descriptive for the `xcb_expose_event_t` type other than `x`. Single variables should be restricted to indexes, loops, but `x` is not descriptive.

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/azframework_xcb_files.cmake:
```diff
@@ -6,6 +6,43 @@
+    else()
+        if(EXISTS ${SPLASH_FILE_RESOLVED_PATH})
+            execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_BINARY_DIR}/AzFramework)
+            execute_process(COMMAND ${CMAKE_COMMAND} -E copy ${SPLASH_FILE_RESOLVED_PATH} ${CMAKE_BINARY_DIR}/AzFramework/Splash.png)
```
Why does splash file need to be copied to a new folder?

Fireronin replied:
It was for organization but I can remove this behaviour

spham-amzn replied:
This is still copying the file to the ${CMAKE_BINARY_DIR}. Is a copy really needed? `xxd` should be able to use the absolute `${SPLASH_FILE_RESOLVED_PATH}` path instead of needing to make this copy.

Fireronin replied:
xdd encodes path to file into the variable name, and for this file generation I need path / filename to be the same when passed to xdd

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/AzFramework/XcbNativeWindow.cpp:
```diff
@@ -219,17 +231,143 @@ namespace AzFramework
+                width,
+                height,
+                XCB_IMAGE_FORMAT_Z_PIXMAP,
+                24,
```
Don't use 'magic' numbers. If `24` is the bit depth, then move this to a `constexpr` like what was done for `bytesPerPixel`.

Fireronin commented:
Ok, changes applied

spham-amzn commented on Code/Framework/AzFramework/Platform/Linux/platform_nativeui_linux.cmake:
```diff
@@ -18,13 +18,16 @@ if (${PAL_TRAIT_LINUX_WINDOW_MANAGER} STREQUAL "xcb")
             3rdParty::X11::xcb_xfixes
             3rdParty::X11::xkbcommon
             3rdParty::X11::xkbcommon_X11
+            xcb_image
```
The `3rdParty::X11::` alias is added by this [function](https://github.com/o3de/o3de/blob/development/cmake/3rdParty/FindX11.cmake). It  basically appends that alias to all of the X11 libraries defined by [cmake](https://cmake.org/cmake/help/latest/module/FindX11.html). 
You should be using `3rdParty::X11:xcb_image` since its defined as an import target, so I don't know whats going on in the Docker script.

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/generate_splash_screen_header.cmake:
```diff
@@ -0,0 +1,34 @@
+            execute_process(COMMAND ${CMAKE_COMMAND} -E copy ${SPLASH_FILE_RESOLVED_PATH} ${CMAKE_BINARY_DIR}/Splash.png)
+            file(REMOVE ${CMAKE_BINARY_DIR}/Splash.h)
+            add_definitions(-DSPLASH_IMAGE_EXISTS)
+            execute_process(COMMAND xxd -i Splash.png Splash.h
```
If `xxd` is not installed, this process will fail. You should handle this situation and present an appropriate warning (and undefined `SPLASH_FILE` in this case as well)

spham-amzn commented on Code/Framework/AzFramework/Platform/Common/Xcb/azframework_xcb_files.cmake:
```diff
@@ -6,6 +6,8 @@
 #
 #
 
+
+
```
Since you refactored out the SPLASH_FILE detection, you can probably revert all changes to this file completely.

Fireronin commented:
Ok this should do the trick, if you will give it a green light I will squash it all

michalpelka commented:
Please rebase and squash, fix DCO.
I also found that you have to have cmake 3.27.0 to configure project. xcb-image was introduced to FindX11.cmake recently: https://github.com/Kitware/CMake/commit/858fd8ece8a5b8787428ba6ad93d63637b226090

spham-amzn commented:
> Please rebase and squash, fix DCO. I also found that you have to have cmake 3.27.0 to configure project. xcb-image was introduced to FindX11.cmake recently: [Kitware/CMake@858fd8e](https://github.com/Kitware/CMake/commit/858fd8ece8a5b8787428ba6ad93d63637b226090)

If this is true, then we either have to update the minimum version of cmake in the [documentation](https://www.docs.o3de.org/docs/welcome-guide/requirements/#linux-cmake) and in the [cmake scripts](https://github.com/o3de/o3de/blob/1f2621241bad1c8eadd477993149dae69ae8bf2f/CMakeLists.txt#L10).

The other alternative is explicitly do a `FindLibrary` for `xcb-image` with the 3rdParty::X11:: prefix in the cmake file for XCB.

Fireronin commented:
This version builds with older cmake

michalpelka commented on Code/Framework/AzFramework/Platform/Linux/platform_nativeui_linux.cmake:
```diff
@@ -18,13 +18,16 @@ if (${PAL_TRAIT_LINUX_WINDOW_MANAGER} STREQUAL "xcb")
             3rdParty::X11::xcb_xfixes
             3rdParty::X11::xkbcommon
             3rdParty::X11::xkbcommon_X11
+            xcb_image
```
I had to apply this to compile it. With that works as expected.

```suggestion
            xcb-image
```

Fireronin commented:
Hi, @spham-amzn 
This pr was reviewed by @michalpelka, and here is updated documentation pr, https://github.com/o3de/o3de.org/pull/2557

michalpelka commented:
Also there is update of build config: https://github.com/o3de/o3de/pull/17925

michalpelka commented:
@spham-amzn thank you for the help on this PR. It would be beneficial for this PR, if you would consider a second review.

spham-amzn reviewed (approved):
I think you will still need to work with sig-build to get the new xcb library on the jenkins build machines for AR or this may not pass. And you may want to send out an impactful change message on discord to broadcast the fact that new packages are needed now to build on Linux

adamdbrw commented:
https://github.com/amzn-changml is there any update on this?

amzn-changml commented:
The required dependencies have been updated. I ran AR, but it appears be failing on Windows. You might want to merge latest in Development

Fireronin commented:
@amzn-changml Hi, do you know what might be the deal with failed tests there are failed tests in Android and windows but both of those mention libraries I haven't even touched and are not even remotely connected with this pr

Fireronin committed:
Working splashScreen

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Initial fixes

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Undoing formating part2

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Undoing formating part3

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Undoing formating part4

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Undoing formating part4

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Undoing formating part4

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Undoing formating part4

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Cleanup

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Cleanup - move function as private

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now generates logo.h based on logo.png, this is then unpacked using lib png, so user can specify logo.png on compile time

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now generates logo.h based on logo.png, this is then unpacked using lib png, so user can specify logo.png on compile time

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now generates logo.h based on logo.png, this is then unpacked using lib png, so user can specify logo.png on compile time. (cleanup)

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now generates logo.h based on logo.png, this is then unpacked using lib png, so user can specify logo.png on compile time. (cleanup 2)

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now generates logo.h based on logo.png, this is then unpacked using lib png, so user can specify logo.png on compile time. (cleanup 3)

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now generates logo.h based on logo.png, this is then unpacked using lib png, so user can specify logo.png on compile time. (cleanup 4)

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now generates logo.h based on logo.png, this is then unpacked using lib png, so user can specify logo.png on compile time. (cleanup 4)

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Removed lambda and logo is now tracked may change this later

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Cleanup of variables

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Now works with new file per project and all is behind ifdef so if file is not present functionality will not be compiled in

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Fancy cmake, and bug fix

Removed todo

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Reduced macro use, dco retrigger

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Applying requested changes

xcb 3rdParty prefix removed

xdd existance check

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Removed prefix

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Update Code/Framework/AzFramework/Platform/Linux/platform_nativeui_linux.cmake

Co-authored-by: Michał Pełka <michal.pelka@robotec.ai>
Signed-off-by: Krzysztof Rymski <nudy148@gmail.com>

Added license

Added copyright license

Signed-off-by: Krzysztof Rymski <krzysztof.rymski@robotec.ai>

Fireronin commented:
Rebased onto development

Fireronin commented on Code/Framework/AzFramework/Platform/Common/Xcb/generate_splash_screen_header.cmake:
```diff
@@ -0,0 +1,50 @@
+# If value is set, check if it exists and generate the header file
+if(DEFINED SPLASH_FILE)
+    # check if xdd is available
+    find_program(XXD xxd)
+    if(NOT XXD)
+        message(WARNING "xxd not found, no splash screen will be displayed"
+                " (install xxd to create splash screen header file)")
+        unset(SPLASH_FILE)
```
Also using xxd is quite simple
